Improving deploy button experience

arraydude opened the pull request:
**Issue:** https://www.notion.so/streamlit/Improve-Core-deploy-button-experience-134acf45228e4ff0a39ede92a9ecc81a

**Description:** 
- Adding more repository information to the initial message
- New backMsg which reloads and resends the initial information without rerunning the script
- New modals for each different scenario specified in the Notion document

<img width="549" alt="Screen Shot 2020-12-04 at 6 35 20 PM" src="https://user-images.githubusercontent.com/631997/101225483-853b9680-365f-11eb-99f7-d1960026dc71.png">
<img width="575" alt="Screen Shot 2020-12-04 at 6 30 38 PM" src="https://user-images.githubusercontent.com/631997/101225460-7e148880-365f-11eb-9eed-11584f545f14.png">
<img width="538" alt="Screen Shot 2020-12-04 at 6 30 58 PM" src="https://user-images.githubusercontent.com/631997/101225462-7fde4c00-365f-11eb-9929-b40eebfc2641.png">
<img width="542" alt="Screen Shot 2020-12-04 at 6 31 27 PM" src="https://user-images.githubusercontent.com/631997/101225465-8076e280-365f-11eb-8b4f-4c909c1d4753.png">
<img width="528" alt="Screen Shot 2020-12-04 at 6 32 44 PM" src="https://user-images.githubusercontent.com/631997/101225467-810f7900-365f-11eb-8f82-13842a5f7562.png">
<img width="529" alt="Screen Shot 2020-12-04 at 6 33 41 PM" src="https://user-images.githubusercontent.com/631997/101225469-8240a600-365f-11eb-8582-4498e312330a.png">








---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

arraydude committed:
Showing basic modals

arraydude committed:
report_session

arraydude committed:
Fixing untracked module

arraydude committed:
Dialogs working with continue anyway

arraydude committed:
lint

arraydude committed:
Try again working

arraydude committed:
pyformat

arraydude committed:
Dialogs unit tests

arraydude committed:
Deploy modals unit test

arraydude committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into deploy-button

# Conflicts:
#	frontend/src/components/core/MainMenu/MainMenu.tsx

arraydude committed:
return type

arraydude committed:
checking if it is not none

arraydude committed:
test push

arraydude committed:
test

arraydude committed:
closing dialog when its ok

arraydude committed:
closing dialog when its ok

arraydude committed:
debug

arraydude committed:
debug

arraydude committed:
fixing detached state

arraydude committed:
lint

arraydude committed:
fixing getDeployAppUrl

arraydude committed:
lint

arraydude committed:
useCallback

tvst commented:
* I always get "No Github repository detected" even when when there is a Git repo.

* This wasn't part of the spec, but I just realized that when we run `streamlit hello` the "deploy" menu item opens a modal saying:
   > Error deploying app.
   > ---
   >
   > No Github repository detected.
   > (...)

   I think the desired behavior would be to show something nice talking about Streamlit Sharing and stuff. But I don't know what that would be. So for now how about we just disable that button for "hello"?
   

* "Sign up for Streamlit for Teams" should be a link to streamlit.io/for-teams
* "Try Again" should be "Try again"
* Inside the modal, the filename should not be formated with **bold** but with `code` instead.

arraydude committed:
Delete SessionState.py

arraydude committed:
Delete example.py

arraydude commented:
Answering to @tvst

> * I always get "No Github repository detected" even when when there is a Git repo.

Have you re-generated the protobufs?, are you using it from local development or building it?


> * This wasn't part of the spec, but I just realized that when we run `streamlit hello` the "deploy" menu item opens a modal saying:
>   > ## Error deploying app.
>   > No Github repository detected.
>   > (...)
>   
>   
>   I think the desired behavior would be to show something nice talking about Streamlit Sharing and stuff. But I don't know 
what that would be. So for now how about we just disable that button for "hello"?

I didn't realize about this , after checking it this is what I'm getting:
<img width="989" alt="Screen Shot 2020-12-08 at 10 02 22 PM" src="https://user-images.githubusercontent.com/631997/101559997-16cd4000-39a1-11eb-98bf-08d871a86208.png">


> * "Sign up for Streamlit for Teams" should be a link to streamlit.io/for-teams

Currently, the option is just "Streamlit for Teams" and it's pointing to `https://streamlit.io/forteams` which ends being redirected to `https://streamlit.io/for-teams`

Just double-checking, should I change these differences? 

> * "Try Again" should be "Try again"

Ok

> * Inside the modal, the filename should not be formated with **bold** but with `code` instead.

Ok

kmcgrady commented on frontend/craco.config.js:
```diff
@@ -25,7 +25,7 @@ module.exports = {
       // to speed up our slow builds.
       // https://github.com/mzgoddard/hard-source-webpack-plugin
-      webpackConfig.plugins.unshift(new HardSourceWebpackPlugin())
+      // webpackConfig.plugins.unshift(new HardSourceWebpackPlugin())
```
I know this has been on our list consider removing. Did something happen that motivated this?

arraydude replied:
Thanks for catching this, I just removed it because I've been having some cache issues regarding this. I'll revert this

arraydude replied:
done

kmcgrady commented on frontend/src/components/core/MainMenu/MainMenu.test.tsx:
```diff
@@ -19,6 +19,17 @@ import React from "react"
+  UncommittedChanges,
+  RepoIsAhead,
+  UntrackedFiles,
+} from "../StreamlitDialog/DeployErrorDialogs"
```
nbd but probably `components/core/StreamlitDialog/DeployErrorDialogs` is preferred. Might as well export the types here, so it all comes from one import.

arraydude replied:
I'll do it. thanks

arraydude replied:
done

kmcgrady commented on frontend/src/components/core/MainMenu/MainMenu.tsx:
```diff
@@ -185,6 +214,82 @@ const MenuListItem = forwardRef<HTMLLIElement, MenuListItemProps>(
+      return
+    }
+
+    if (isDeployErrorModalOpen) closeDialog()
```
This line is a little weird, when clicking deploy, it closes a dialog. Any reason why this exists? Probably deserves a comment if it's good. Less important, but I don't see us using the one liners for if statements, so I'd probably do the full fledged if with the block.

arraydude replied:
I have to do because the state of the dialog is being handled on App, so in the case that you have a dialog , and then you solve the problem and then "Try Again" the dialog will close and use `getDeployAppUrl` to do the open.window with the correct coordinates

arraydude replied:
I'll add a comment

arraydude replied:
done

kmcgrady commented on lib/streamlit/cursor.py:
```diff
@@ -152,7 +152,7 @@ def __init__(
         parent_path: Tuple[int, ...] = (),
         index: int = 0,
-        **props
+        **props,
```
Seeing a bunch of formatting, which is weird...have you updated the latest pipfile or something...wonder if the formatter has not been updated. Want to make sure we aren't just reverting cause the styles are not a problem.

arraydude replied:
This was done by the formatter, just in case I'll run it again after fixing all the changes with the latest from develop

kmcgrady commented on lib/streamlit/report_session.py:
```diff
@@ -350,28 +350,52 @@ def get_deploy_params(self):
+            }
         except:
             # Issues can arise based on the git structure
             # (e.g. if branch is in DETACHED HEAD state,
```
This comment probably doesn't fully apply cause we support the detached head state.

arraydude replied:
Yeah! that's true , thank you Ken

arraydude replied:
done

kmcgrady commented on lib/streamlit/report_session.py:
```diff
@@ -350,28 +350,52 @@ def get_deploy_params(self):
+            msg.new_report.deploy_params.uncommitted_files[:] = deploy_params[
+                "uncommitted_files"
+            ]
+            msg.new_report.deploy_params.is_ahead = deploy_params["is_ahead"]
```
Wonder if there's a way to simplify this code, and just have at the end.

`msg.new_report.deploy_params = X`. Just saves really setting all the values individually.

kmcgrady replied:
I think you can import `DeployParams` and then set those accordingly.

arraydude replied:
Do you mean sending the 'modal state' directly from the python side ? instead of handling it in javascript ?

kmcgrady replied:
Sorry no, I meant like.

```python
from streamlit.proto.NewReport_pb2 import NewReport, DeployParams

#...

new_report = NewReport()
new_report.report_id = self._report.report_id
#...

dep_params = DeployParams()
# ...
if deploy_params["info"] is not None:
    repo, branch, module = deploy_params["info"]
    dep_params.repository = repo
    dep_params.branch = branch	
    dep_params.module = module

# ...and so on....

new_report.deploy_params = dep_params
msg.new_report = new_report
```

It saves from all the `msg.new_report.deploy_params` mentions that might make the code look cleaner then.

kmcgrady commented on lib/streamlit/report_session.py:
```diff
@@ -423,6 +447,10 @@ def _enqueue_report_finished_message(self, status):
         msg.report_finished = status
         self.enqueue(msg)
 
+    def handle_reload_report_message(self):
```
Making sure I understand. The main purpose is to ensure that we can not generate a new id when rerunning the report. What's the reasoning behind this?

arraydude replied:
Prevent a rerun, we are only reloading the initial messages ... It was a requirement, prevent a rerun when you `Try Again`

kmcgrady replied:
Gotcha. I understand why you would be doing that. I just want to make sure that's what we should do. I'll circle back here.

kmcgrady commented on lib/streamlit/git_util.py:
```diff
@@ -106,3 +119,21 @@ def get_repo_info(self):
+        if not self.is_valid():
+            return None
+
+        return [item.a_path for item in self.repo.index.diff(None)]
```
I imagine this will raise questions about the amount of times we make calls to git. Any chance to bring in an option to disable all of these? Probably a question to include @tvst 

https://github.com/streamlit/streamlit/issues/2323#issuecomment-727994990

kmcgrady replied:
Following up on this comment.

kmcgrady commented on lib/streamlit/git_util.py:
```diff
@@ -46,13 +46,26 @@ def is_valid(self) -> bool:
+
+    @property
+    def is_head_detached(self):
+        return self.repo.head.is_detached
```
Not a big deal since we likely check it multiple times but we should check if the repo is valid.

arraydude replied:
Makes sense, I'll do it

arraydude replied:
As we are getting the repo, branch, and module from remote. when there's a detached state we don't have that information so the is valid method will always fail in that case

kmcgrady commented on frontend/src/components/core/MainMenu/MainMenu.tsx:
```diff
@@ -185,6 +214,82 @@ const MenuListItem = forwardRef<HTMLLIElement, MenuListItemProps>(
+      return
+    }
+
+    if (deployParams?.isHeadDetached) {
```
Curious if typescript is requiring the `?` cause you do check it does not exist in 225.

Think it might be nice since this is consistent to just do:
`const { isHeadDetached, module, untrackedFiles, uncommittedFiles, isAhead } = deployParams`. Saves a lot of `deployParams` checks.

There's probably a more clever way to make this easy, but I figure it is probably fine to just be explicit.

arraydude replied:
done

kmcgrady reviewed (commented):
Sending thoughts on code. Will test more tomorrow. Most of it is a lot of answering questions and some areas to improve stylistically.

Thanks for the work! It's very thorough!

kmcgrady commented on frontend/src/components/core/StreamlitDialog/styled-components.ts:
```diff
@@ -27,3 +27,9 @@ export const StyledShortcutLabel = styled.span(({ theme }) => ({
+  "& > ul": {
+    paddingLeft: "1rem",
+  },
+}))
```
Love seeing this used. I wonder if we should just increase indentation on `ul`s but not a big problem right now.

arraydude replied:
Yeah! , you did a great job with these styled-components Ken, I'll add a bit more space.

arraydude replied:
done

arraydude committed:
Tvst comments:
- "Sign up for Streamlit for Teams" should be a link to streamlit.io/for-teams
- "Try Again" should be "Try again"
- Inside the modal, the filename should not be formated with bold but with code instead.

arraydude committed:
Merge remote-tracking branch 'origin/deploy-button' into deploy-button

arraydude committed:
Revert HardSourceWebpackPlugin change

arraydude committed:
using alias instead of relative path, and comments

arraydude committed:
destructuring deployParams

arraydude committed:
more padding

arraydude committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into deploy-button

arraydude committed:
make pyformat

arraydude committed:
make pyformat

randyzwitch reviewed (approved):
Not clear to me where I'm a code owner, but I'll approve since Ken appears to be giving this a detailed lookover

arraydude committed:
Fixing uncommited files and prevent showing deploy button when `streamlit hello`

tvst commented:
I installed the latest wheel file from @arraydude and this still doesn't work. Here's what I see in the logs:

```
$ streamlit run streamlit_app.py
=== git path streamlit_app.py
=== git path /media/storage/Projects/streamlit/test_app/streamlit_app.py
=== git version (2, 28, 0)
=== git is minimum
=== git version (2, 28, 0)
=== git is minimum
=== git repo <git.repo.base.Repo '/media/storage/Projects/streamlit/test_app/.git'>

  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://192.168.1.253:8501

=== git repo <git.repo.base.Repo '/media/storage/Projects/streamlit/test_app/.git'>
=== repo <streamlit.git_util.GitRepo object at 0x7f147cda9550>
=== info None
=== detached False
=== untracked ['Pipfile', 'secrets.py']
=== uncommitted []
=== debugg []
=== ahead False


CLICKED ON "DEPLOY" MENU ITEM:
(no new logs)

CLICKED ON "TRY AGAIN":
=== git path /media/storage/Projects/streamlit/test_app/streamlit_app.py
=== git version (2, 28, 0)
=== git is minimum
=== git repo <git.repo.base.Repo '/media/storage/Projects/streamlit/test_app/.git'>
=== repo <streamlit.git_util.GitRepo object at 0x7f147ccc9160>
=== info None
=== detached False
=== untracked ['Pipfile', 'secrets.py']
=== uncommitted []
=== debugg []
=== ahead False
```

You can repro by downloading the wheel file from Emiliano and then creating this Dockerfile on that same folder:
```
FROM python:3.7-slim

WORKDIR /home

ADD streamlit-0.72.0-py2.py3-none-any.whl .
RUN pip install streamlit-0.72.0-py2.py3-none-any.whl

RUN apt update && apt-get install -y git && rm -rf /var/lib/apt/lists/*

RUN echo '\n\
import streamlit as st\n\
st.write("hello")\n\
' > streamlit_app.py

RUN git config --global user.email "demo@streamlit.io"
RUN git config --global user.name "Demo User"

RUN git init
RUN git add streamlit_app.py
RUN git commit -m "First commit"

CMD ["streamlit", "run", "streamlit_app.py"]
```

Then just call this as usual:
```
docker build --tag=testapp:1 .
docker run -p 8501:8501 testapp:1
```

kmcgrady commented on frontend/src/components/core/MainMenu/MainMenu.tsx:
```diff
@@ -185,6 +215,92 @@ const MenuListItem = forwardRef<HTMLLIElement, MenuListItemProps>(
+
+      showDeployError(dialog.title, dialog.body)
+
+      return
```
Not a big deal, but you can probably just put the return all on one line.

`return showDeplyError(...)` for each one.

The result is exactly the same, but the code is a little tighter. I'm assuming the newlines are linting errors/formatter settings.

kmcgrady commented on lib/streamlit/git_util.py:
```diff
@@ -106,3 +119,24 @@ def get_repo_info(self):
+
+            return list(
+                self.repo.iter_commits(
+                    "/".join([remote.name, branch_name]) + ".." + branch_name
```
Another small nit, but I'd probably just create a variable like `remote_branch = "/".join(...)`

You can also use python [formatted strings](https://realpython.com/python-formatted-output/#the-python-formatted-string-literal-f-string) It makes this really clean.

`f'{remote_branch}..{branch_name}'`

arraydude committed:
Fixing wordings and unit test

arraydude committed:
using formatted strings

arraydude committed:
using formatted strings

arraydude committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into deploy-button

# Conflicts:
#	lib/streamlit/elements/data_frame_proto.py
#	lib/streamlit/elements/time_widgets.py
#	lib/streamlit/elements/vega_lite.py
#	lib/streamlit/elements/write.py

arraydude committed:
Deprecating deployParams from initialMessage and adding a new one

arraydude committed:
Updating frontend tests

arraydude committed:
removing python tests

arraydude committed:
Using IGitInfo instead of GitInfo

arraydude committed:
Merge pull request #19 from arraydude/deploy-button-v2

Deprecating deployParams from initialMessage and adding a new one

arraydude committed:
Removing console.log

arraydude committed:
Removing unused method

arraydude committed:
snap

arraydude committed:
cleaning

arraydude committed:
load_git_info

akrolsmir commented:
Closing this since we've moved forward with #2552 and #2535 ~
